src/python/pants/engine/legacy/parser.py

```diff
@@ -125,5 +130,35 @@ def parse(self, filepath, filecontent):
     # _intentional_ mutation would require a deep clone, which doesn't seem worth the cost at
     # this juncture.
     self._parse_context._storage.clear(os.path.dirname(filepath))
-    six.exec_(python, dict(self._symbols))
+
+    symbols = dict(self._symbols)
+    if self._build_file_import_behavior != 'allow':
+      # This is not secure sandboxing, because people could replace __import__ again themselves
+      # if they wanted (and there are plenty of other ways to escape a python "sandbox"), but it
+      # should be sufficient to tell the casual user that they're doing something wrong.
+      builtins = dict(__builtins__)
+      if self._build_file_import_behavior == 'warn':
+        import_hook = lambda import_name, *args: _warn_on_import(__import__, filepath, import_name, *args)
+      elif self._build_file_import_behavior == 'error':
+        import_hook = _fail_on_import
+      else:
+        raise ParseError("Didn't know what to do for build_file_import_behavior value {}".format(
+          self._build_file_import_behavior
+        ))
+      builtins['__import__'] = import_hook
+      symbols['__builtins__'] = builtins
+
+    six.exec_(python, symbols)
     return list(self._parse_context._storage.objects)
+
+
+def _warn_on_import(import_builtin, filepath, import_name, *args):
+  logger = logging.getLogger(__name__)
+  logger.warn('BUILD file at {} tried to import {} - import statements should be avoided'.format(
+    filepath, import_name
+  ))
+  return import_builtin(import_name, *args)
+
+
+def _fail_on_import(name, *args):
```

Comment:
This error message is less useful than the other one because it doesn't include the bad file path.

Reply:
The tests assert that the file name is included in the error message.

```
$ ./pants --build-file-imports=error --print-exception-stacktrace=false run testprojects/src/python/build_file_imports_module:hello
Exception caught: (<class 'pants.build_graph.address_lookup_error.AddressLookupError'>)

Exception message: Build graph construction failed: ExecutionError Multiple exceptions encountered:
  MappingError: Failed to parse testprojects/src/python/build_file_imports_module/BUILD:
import statements have been banned, but tried to import os.path
  MappingError: Failed to parse testprojects/src/python/build_file_imports_module/BUILD:
import statements have been banned, but tried to import os.path
```

(Any idea why it's double-printing?)

Reply:
I think because there were multiple roots, and they failed with the same error. Needs work. I'm going to try and convince Emily to work on https://github.com/pantsbuild/pants/issues/3695 soon, heh.